tests/backend.cpp

```diff
@@ -382,6 +383,188 @@ TEST_F(backendTest, atomicFetchAddPointer) {
+		for(std::size_t i=0; i<array_size; i++){
+			tmp = array[i];
+		}
+		ASSERT_LE(tmp, max_total);
+	}
+
+	// Wait for the flag change to be visible on every node
+	while(*flag != 1){
+		ASSERT_LT(std::chrono::system_clock::now(), max_time);
+		argo::backend::selective_acquire(flag, sizeof(unsigned));
+	}
+
+	// Check the array values on every node
+	argo::backend::selective_acquire(array, array_size*sizeof(int));
+	int count = 0;
+	const int expected = i_const*array_size;
+	for(std::size_t i=0; i<array_size; i++){
+		count += array[i];
+	}
+	ASSERT_EQ(count, expected);
+
+	// Clean up
+	argo::codelete_array(array);
+}
+
+/**
+ * @brief Test selective coherence on unaligned acquires and releases
+ */
+TEST_F(backendTest, selectiveUnaligned) {
+	const std::size_t array_size = 2097152;
+	const std::size_t ua_chunk_size = 256;
+	int* array = argo::conew_array<int>(array_size);
+	unsigned int* flag(argo::conew_<unsigned>(0));
+	std::chrono::system_clock::time_point max_time =
+		std::chrono::system_clock::now() + deadlock_threshold;
+
+	// Initialize
+	if(argo::node_id() == 0){
+		for(std::size_t i=0; i<array_size; i++){
+			array[i] = 0;
+		}
+	}
+	argo::barrier();
+
+	// Set array elements on node 0, then set flag
+	if(argo::node_id() == 0){
+		// Write an unaligned chunk crossing a (remote node) boundary
+		for(std::size_t i=ua_chunk_size*7231; i<ua_chunk_size*7233; i++){
+			array[i] = i_const;
+		}
+		argo::backend::selective_release(&array[ua_chunk_size*7231],
+				(ua_chunk_size*2)*sizeof(int));
+
+		*flag = 1;
+		argo::backend::selective_release(flag, sizeof(unsigned));
+	}
+	// Read the set values on every other node to make sure that some
+	// version of the pages involved are already cached
+	else{
+		int tmp = 0;
+		const int max_total = i_const*ua_chunk_size*2;
+		for(std::size_t i=ua_chunk_size*7231; i<ua_chunk_size*7233; i++){
+			tmp = array[i];
+		}
+		ASSERT_LE(tmp, max_total);
+	}
+
+	// Wait for the flag change to be visible on every node
+	while(*flag != 1){
+		ASSERT_LT(std::chrono::system_clock::now(), max_time);
+		argo::backend::selective_acquire(flag, sizeof(unsigned));
+	}
+
+	// Check the set array values on every node
+	argo::backend::selective_acquire(&array[ua_chunk_size*7231],
+			(ua_chunk_size*2)*sizeof(int));
+	int count = 0;
+	const int expected = i_const*ua_chunk_size*2;
+	for(std::size_t i=0; i<array_size; i++){
+		count += array[i];
+	}
+	ASSERT_EQ(count, expected);
+
+	// Clean up
+	argo::codelete_array(array);
+}
+
+/**
+ * @brief Test write buffer under load with random access patterns
+ */
+TEST_F(backendTest, writeBufferLoad) {
+	const std::size_t array_size = 4000000; // Just under max size 16Mb
+	const std::size_t num_writes = array_size/20; // Not too many random writes
+	int* array = argo::conew_array<int>(array_size);
+
+	// Random device to ensure accesses are irregular in order to expose
+	// random ordering between writebacks to different nodes
+	std::random_device rd;
+	std::mt19937 rng(rd());
```

Comment:
Question for @kostis
Is it okay to have a unit test that produces a truly random sequence, or should the seed here be made a constant to make sure all encountered bugs are perfectly reproducible?

Reply:
If the outcome of the test is deterministic, it is OK, I think.

Reply:
I have not changed this test in the latest push. If this is not urgent perhaps it should be left as an issue (Ensure deterministic results in backend test).